docs/proposals/llm-hyperparameters-tuning-api.md

```diff
@@ -0,0 +1,135 @@
+# LLM Hyperparameters Tuning API Proposal
+
+- [LLM Hyperparameters Tuning API Proposal](#llm-hyperparameters-tuning-api-proposal)
+  * [Links](#links)
+  * [Motivation](#motivation)
+  * [Goal](#goal)
+  * [Design for API](#design-for-api)
+    + [Example](#example)
+  * [Implementation](#implementation)
+  * [Advanced Functionalities](#advanced-functionalities)
+
+## Links
+
+- [katib/issues#2291 (Tuning API in Katib for LLMs)](https://github.com/kubeflow/katib/issues/2291)
+
+## Motivation
+
+LLMs are experiencing rapid growth in popularity and evolution, leading to an increased demand for LLMOps in Kubernetes environment. Hyperparameter tuning can greatly improve model performance but is often labor-intensive for humans. We want to provide an LLM hyperparameters tuning API which automates the process and is user-friendly for data scientists.
+
+## Goal
+
+We aim to develop a high-level API for tuning LLM hyperparameters, which abstracts away infrastructure complexities for data scientists, and facilitates easy hyperparameter tuning by enabling the seamless import of pretrained models and datasets from platforms like HuggingFace and S3, thereby improving accessibility and ease of use.
+
+## Design for API
+
+![Design for API](https://github.com/helenxie-bit/kubeflow/blob/1cdff9f77d68df824bea128e419dc612e172173b/design_api.jpg)
+
+```python
+# Reuse existing assets in Training Operator for importing models and datasets
+import kubeflow.storage_initializer.hugging_face
+import kubeflow.storage_initializer.s3
+
+from kubeflow import katib
+katib_client = katib.KatibClient()
+
+# Arguments related to the model provider including credentials
+model_args = modelProviderClass()
+
+# Arguments related to the dataset provider including credentials 
+dataset_args = datasetProviderClass()
+
+# Define hyperparameters and search space
+parameters = {
+	"learning_rate": katib.search.double(min=[], max=[]),
+	"batch_size": katib.search.int(min=[], max=[]),
+	"num_epoch": katib.search.int(min=[], max=[]),
+	"weight_decay": katib.search.double(min=[], max=[])
+	...
+}
+
+# Create experiment for tuning
+exp_name = "llm-tune-test"
+katib_client.tune(
+	name=exp_name,
+	model=model_args,
+	dataset=dataset_args,
```

Comment:
Should we be consistent with `train` API and introduce `model_provider_parameters`, `dataset_provider_parameters`, `trainer_parameters` ?
I think, we should also add `pytorch_parameters` to set `num_workers` and `num_procs_per_worker`.

Any thoughts what should be the `objective` in case of LLM HyperParameter optimization during fine-tuning ?

Reply:
Should parallel_trial_count be same as num_workers?

Reply:
No, since one Trial is one PyTorchJob, and one PyTorchJob worker is one Kubernetes Pod.
E.g. Trial -> PyTorchJob -> Pods
In Katib PyTorchJob is just Trial worker: https://www.kubeflow.org/docs/components/katib/reference/architecture/#katib-architecture-for-hyperparameter-tuning

Reply:
I have renamed `model` and `dataset` to `model_provider_parameters` and `dataset_provider_parameters`, respectively. Additionally, I introduced `trainer_parameters` to keep consistent with the `train` API.

I have also added PyTorch_parameters including `num_workers`, `num_procs_per_worker`, and `resources_per_worker`.

Regarding the `objective` function, please refer to the example in my latest proposal. I understand that the `objective` function should be a callable that returns the metrics we aim to optimize. I have integrated the `train` API to streamline this process. In the `objective` function, the `train` API handles downloading pretrained models and datasets, performing the training, and returning results that typically include metrics such as `eval_accuracy` and `eval_loss`, assuming Huggingface’s `trainer.train()` is utilized. Then we can use these results to extract the objective metric for hyperparameter optimization.

Reply:
> No, since one Trial is one PyTorchJob, and one PyTorchJob worker is one Kubernetes Pod. E.g. Trial -> PyTorchJob -> Pods In Katib PyTorchJob is just Trial worker: https://www.kubeflow.org/docs/components/katib/reference/architecture/#katib-architecture-for-hyperparameter-tuning

Thank you for the explanation! It seems users can configure two sets of resource parameters: those for the `trial` (such as `max_trial_count`, `max_failed_trial_count`, `parallel_trial_count`, `resources_per_trial`) and those for the `PyTorchJob` (like `num_workers`, `num_procs_per_worker`, `resources_per_worker`). Is that correct?

I'm wondering if there is any relationship between the values of these parameters. For instance, if users want to set both `resources_per_trial` and `resources_per_worker`, should the values in `resources_per_trial`—such as the number of GPUs—be calculated as the number of GPUs set in `resources_per_worker` multiplied by `num_workers`?

Reply:
> It seems users can configure two sets of resource parameters

Yes, that is correct. The flow looks like this: Trial -> PyTorchJob -> Pod.
In case of Katib `resources_per_trial` = `resources_per_worker`, that is why we can keep just `resources_per_trial` parameter.

@johnugeorge @tenzen-y @helenxie-bit What do we think about this ? Introduce `pytorch_config` parameter  in tune where we can set this:
```python
b = tune(
    name="llm-tune",
    pytorch_config=katib.PyTorchConfig(
        num_workers=5,
        num_procs_per_worker=2
    ),
    ...
)
```

Reply:
Introducing `PyTorch_config` parameter can helper users better differentiate between the parameters related to trials and PyTorchJob workers. It makes sense for me.